libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,863 @@
+
+
+/**
+ * Used to return the appropriate representation of a case insensitive string for the EOSIO Key Value database.
+ *
+ * @param val - The string to be made case-insensitive
+ * @return The binary representation of the case-insensitive string
+ */
+inline key_type make_insensitive(const std::string& val) {
+   return make_key(val, true);
+}
+
+
+/**
+ * @defgroup keyvalue Key Value Table
+ * @ingroup contracts
+ *
+ * @brief Defines an EOSIO Key Value Table
+ * @details EOSIO Key Value API provides a C++ interface to the EOSIO Key Value database.
+ * Key Value Tables require 1 primary index, of any type that can be serialized to a binary representation.
+ * Key Value Tables support 0 or more secondary index, of any type that can be serialized to a binary representation.
+ * Indexes must be a member variable or a member function.
+ *
+ * @tparam Class     - the name of the class of the user defined table that inherits from eosio::kv_table
+ * @tparam T         - the type of the data stored as the value of the table
+ * @tparam TableName - the name of the table
+ * @tparam DbName    - the type of the EOSIO Key Value database. Defaulted to eosio.kvram
+  */
+template<typename Class, typename T, eosio::name::raw TableName, eosio::name::raw DbName = eosio::name{"eosio.kvram"}>
+class kv_table {
+
+   enum class kv_it_stat {
+      iterator_ok     = 0,  // Iterator is positioned at a key-value pair
+      iterator_erased = -1, // The key-value pair that the iterator used to be positioned at was erased
+      iterator_end    = -2, // Iterator is out-of-bounds
+   };
+
+public:
+   /**
+    * @ingroup keyvalue
+    *
+    * @brief Defines an index on an EOSIO Key Value Table
+    * @details A Key Value Index allows a user of the table to search based on a given field.
+    * The only restrictions on that field are that it is serializable to a binary representation.
+    * Convenience functions exist to handle most of the primitive types as well as some more complex types, and are
+    * used automatically where possible.
+    */
+   class kv_index {
+      class iterator {
+      public:
+         iterator(eosio::name contract_name, uint32_t itr, kv_it_stat itr_stat, kv_index* idx) :
+                  contract_name{contract_name}, itr{itr}, itr_stat{itr_stat}, idx{idx} {}
+
+         /**
+          * Returns the value that the iterator points to.
+          * @ingroup keyvalue
+          *
+          * @return The value that the iterator points to.
+          */
+         T value() const {
+            using namespace detail;
+
+            eosio::check(itr_stat != kv_it_stat::iterator_end, "Cannot read end iterator");
+
+            uint32_t value_size;
+            uint32_t actual_value_size;
+            uint32_t offset = 0;
+
+            // call once to get the value_size
+            internal_use_do_not_use::kv_it_value(itr, 0, (char*)nullptr, 0, value_size);
+
+            eosio::check(value_size > 0, "Cannot read a value of size 0");
```

Comment:
Why not?

Reply:
For the purposes of serialization/deserialization, what would a size 0 value actually mean? Don't all types have at least size 1?

Reply:
An empty struct serializes to size 0. This is not the sizeof operator.